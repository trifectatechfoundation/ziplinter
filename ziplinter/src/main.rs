use rc_zip_sync::ReadZip;
use std::ops::Deref;

fn main() {
    let mut args = std::env::args();
    let _ = args.next();
    let path = args.next().unwrap();

    let file = std::fs::File::open(path).unwrap();
    let archive = file.read_zip().unwrap();

    eprintln!(
        "{}",
        serde_json::to_string_pretty(&archive.deref()).unwrap()
    )
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::assert_json_snapshot;
    use std::{error::Error, path::Path};

    fn process_zip_file(zip_path: &Path) -> Result<serde_json::Value, Box<dyn Error>> {
        let file = std::fs::File::open(zip_path).unwrap();
        let archive = file.read_zip();

        Ok(serde_json::to_value(archive?.deref())?)
    }

    #[test]
    fn snapshot_zip_files() {
        let fixtures_dir = std::env::current_dir()
            .unwrap()
            .join("../testdata")
            .canonicalize()
            .unwrap();

        println!("fixtures_dir: {}", fixtures_dir.display());

        for entry in std::fs::read_dir(fixtures_dir).expect("Failed to read fixtures directory") {
            let entry = entry.expect("Failed to read entry");
            let path = entry.path();

            if path
                .extension()
                .is_some_and(|e| e.eq_ignore_ascii_case("zip"))
            {
                let Ok(result) = process_zip_file(&path) else {
                    continue;
                };
                println!(
                    "current file: {}",
                    Path::new(path.file_name().unwrap()).display()
                );
                assert_json_snapshot!(
                    format!("{}", Path::new(path.file_name().unwrap()).display()),
                    result
                );
            }
        }
    }
}
