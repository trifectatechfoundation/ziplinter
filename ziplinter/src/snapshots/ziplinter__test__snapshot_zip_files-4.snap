---
source: ziplinter/src/main.rs
expression: result
---
{
  "comment": "This is the source code repository for the Go programming language.  \r\n\r\nFor documentation about how to install and use Go,\r\nvisit https://golang.org/ or load doc/install.html in your web browser.\r\n\r\nAfter installing Go, you can view a nicely formatted\r\ndoc/install.html by running godoc --http=:6060\r\nand then visiting http://localhost:6060/doc/install.html.\r\n\r\nUnless otherwise noted, the Go source files are distributed\r\nunder the BSD-style license found in the LICENSE file.\r\n\r\n--\r\n\r\nBinary Distribution Notes\r\n\r\nIf you have just untarred a binary Go distribution, you need to set\r\nthe environment variable $GOROOT to the full path of the go\r\ndirectory (the one containing this README).  You can omit the\r\nvariable if you unpack it into /usr/local/go, or if you rebuild\r\nfrom sources by running all.bash (see doc/install.html).\r\nYou should also add the Go binary directory $GOROOT/bin\r\nto your shell's path.\r\n\r\nFor example, if you extracted the tar file into $HOME/go, you might\r\nput the following in your .profile:\r\n\r\n    export GOROOT=$HOME/go\r\n    export PATH=$PATH:$GOROOT/bin\r\n\r\nSee doc/install.html for more details",
  "directory_headers": [
    {
      "comment": "",
      "compressed_size": 598,
      "crc32": 2181565152,
      "creator_version": {
        "host_system": "Unix",
        "version": 30
      },
      "disk_nbr_start": 0,
      "external_attrs": 2175008768,
      "extra": [
        85,
        84,
        5,
        0,
        3,
        75,
        201,
        127,
        76,
        117,
        120,
        11,
        0,
        1,
        4,
        232,
        3,
        0,
        0,
        4,
        232,
        3,
        0,
        0
      ],
      "flags": 0,
      "header_offset": 0,
      "internal_attrs": 1,
      "method": "Deflate",
      "modified": {
        "date": 15650,
        "time": 24352
      },
      "name": "README",
      "reader_version": {
        "host_system": "MsDos",
        "version": 20
      },
      "uncompressed_size": 1096
    }
  ],
  "encoding": "Utf8",
  "entries": [
    {
      "accessed": null,
      "comment": "",
      "compressed_size": 598,
      "crc32": 2181565152,
      "created": null,
      "flags": 0,
      "gid": 1000,
      "header_offset": 0,
      "method": "Deflate",
      "mode": 420,
      "modified": "2010-09-02T15:56:59Z",
      "name": "README",
      "reader_version": {
        "host_system": "MsDos",
        "version": 20
      },
      "uid": 1000,
      "uncompressed_size": 1096
    }
  ],
  "eocd": {
    "dir": {
      "inner": {
        "comment": [
          84,
          104,
          105,
          115,
          32,
          105,
          115,
          32,
          116,
          104,
          101,
          32,
          115,
          111,
          117,
          114,
          99,
          101,
          32,
          99,
          111,
          100,
          101,
          32,
          114,
          101,
          112,
          111,
          115,
          105,
          116,
          111,
          114,
          121,
          32,
          102,
          111,
          114,
          32,
          116,
          104,
          101,
          32,
          71,
          111,
          32,
          112,
          114,
          111,
          103,
          114,
          97,
          109,
          109,
          105,
          110,
          103,
          32,
          108,
          97,
          110,
          103,
          117,
          97,
          103,
          101,
          46,
          32,
          32,
          13,
          10,
          13,
          10,
          70,
          111,
          114,
          32,
          100,
          111,
          99,
          117,
          109,
          101,
          110,
          116,
          97,
          116,
          105,
          111,
          110,
          32,
          97,
          98,
          111,
          117,
          116,
          32,
          104,
          111,
          119,
          32,
          116,
          111,
          32,
          105,
          110,
          115,
          116,
          97,
          108,
          108,
          32,
          97,
          110,
          100,
          32,
          117,
          115,
          101,
          32,
          71,
          111,
          44,
          13,
          10,
          118,
          105,
          115,
          105,
          116,
          32,
          104,
          116,
          116,
          112,
          115,
          58,
          47,
          47,
          103,
          111,
          108,
          97,
          110,
          103,
          46,
          111,
          114,
          103,
          47,
          32,
          111,
          114,
          32,
          108,
          111,
          97,
          100,
          32,
          100,
          111,
          99,
          47,
          105,
          110,
          115,
          116,
          97,
          108,
          108,
          46,
          104,
          116,
          109,
          108,
          32,
          105,
          110,
          32,
          121,
          111,
          117,
          114,
          32,
          119,
          101,
          98,
          32,
          98,
          114,
          111,
          119,
          115,
          101,
          114,
          46,
          13,
          10,
          13,
          10,
          65,
          102,
          116,
          101,
          114,
          32,
          105,
          110,
          115,
          116,
          97,
          108,
          108,
          105,
          110,
          103,
          32,
          71,
          111,
          44,
          32,
          121,
          111,
          117,
          32,
          99,
          97,
          110,
          32,
          118,
          105,
          101,
          119,
          32,
          97,
          32,
          110,
          105,
          99,
          101,
          108,
          121,
          32,
          102,
          111,
          114,
          109,
          97,
          116,
          116,
          101,
          100,
          13,
          10,
          100,
          111,
          99,
          47,
          105,
          110,
          115,
          116,
          97,
          108,
          108,
          46,
          104,
          116,
          109,
          108,
          32,
          98,
          121,
          32,
          114,
          117,
          110,
          110,
          105,
          110,
          103,
          32,
          103,
          111,
          100,
          111,
          99,
          32,
          45,
          45,
          104,
          116,
          116,
          112,
          61,
          58,
          54,
          48,
          54,
          48,
          13,
          10,
          97,
          110,
          100,
          32,
          116,
          104,
          101,
          110,
          32,
          118,
          105,
          115,
          105,
          116,
          105,
          110,
          103,
          32,
          104,
          116,
          116,
          112,
          58,
          47,
          47,
          108,
          111,
          99,
          97,
          108,
          104,
          111,
          115,
          116,
          58,
          54,
          48,
          54,
          48,
          47,
          100,
          111,
          99,
          47,
          105,
          110,
          115,
          116,
          97,
          108,
          108,
          46,
          104,
          116,
          109,
          108,
          46,
          13,
          10,
          13,
          10,
          85,
          110,
          108,
          101,
          115,
          115,
          32,
          111,
          116,
          104,
          101,
          114,
          119,
          105,
          115,
          101,
          32,
          110,
          111,
          116,
          101,
          100,
          44,
          32,
          116,
          104,
          101,
          32,
          71,
          111,
          32,
          115,
          111,
          117,
          114,
          99,
          101,
          32,
          102,
          105,
          108,
          101,
          115,
          32,
          97,
          114,
          101,
          32,
          100,
          105,
          115,
          116,
          114,
          105,
          98,
          117,
          116,
          101,
          100,
          13,
          10,
          117,
          110,
          100,
          101,
          114,
          32,
          116,
          104,
          101,
          32,
          66,
          83,
          68,
          45,
          115,
          116,
          121,
          108,
          101,
          32,
          108,
          105,
          99,
          101,
          110,
          115,
          101,
          32,
          102,
          111,
          117,
          110,
          100,
          32,
          105,
          110,
          32,
          116,
          104,
          101,
          32,
          76,
          73,
          67,
          69,
          78,
          83,
          69,
          32,
          102,
          105,
          108,
          101,
          46,
          13,
          10,
          13,
          10,
          45,
          45,
          13,
          10,
          13,
          10,
          66,
          105,
          110,
          97,
          114,
          121,
          32,
          68,
          105,
          115,
          116,
          114,
          105,
          98,
          117,
          116,
          105,
          111,
          110,
          32,
          78,
          111,
          116,
          101,
          115,
          13,
          10,
          13,
          10,
          73,
          102,
          32,
          121,
          111,
          117,
          32,
          104,
          97,
          118,
          101,
          32,
          106,
          117,
          115,
          116,
          32,
          117,
          110,
          116,
          97,
          114,
          114,
          101,
          100,
          32,
          97,
          32,
          98,
          105,
          110,
          97,
          114,
          121,
          32,
          71,
          111,
          32,
          100,
          105,
          115,
          116,
          114,
          105,
          98,
          117,
          116,
          105,
          111,
          110,
          44,
          32,
          121,
          111,
          117,
          32,
          110,
          101,
          101,
          100,
          32,
          116,
          111,
          32,
          115,
          101,
          116,
          13,
          10,
          116,
          104,
          101,
          32,
          101,
          110,
          118,
          105,
          114,
          111,
          110,
          109,
          101,
          110,
          116,
          32,
          118,
          97,
          114,
          105,
          97,
          98,
          108,
          101,
          32,
          36,
          71,
          79,
          82,
          79,
          79,
          84,
          32,
          116,
          111,
          32,
          116,
          104,
          101,
          32,
          102,
          117,
          108,
          108,
          32,
          112,
          97,
          116,
          104,
          32,
          111,
          102,
          32,
          116,
          104,
          101,
          32,
          103,
          111,
          13,
          10,
          100,
          105,
          114,
          101,
          99,
          116,
          111,
          114,
          121,
          32,
          40,
          116,
          104,
          101,
          32,
          111,
          110,
          101,
          32,
          99,
          111,
          110,
          116,
          97,
          105,
          110,
          105,
          110,
          103,
          32,
          116,
          104,
          105,
          115,
          32,
          82,
          69,
          65,
          68,
          77,
          69,
          41,
          46,
          32,
          32,
          89,
          111,
          117,
          32,
          99,
          97,
          110,
          32,
          111,
          109,
          105,
          116,
          32,
          116,
          104,
          101,
          13,
          10,
          118,
          97,
          114,
          105,
          97,
          98,
          108,
          101,
          32,
          105,
          102,
          32,
          121,
          111,
          117,
          32,
          117,
          110,
          112,
          97,
          99,
          107,
          32,
          105,
          116,
          32,
          105,
          110,
          116,
          111,
          32,
          47,
          117,
          115,
          114,
          47,
          108,
          111,
          99,
          97,
          108,
          47,
          103,
          111,
          44,
          32,
          111,
          114,
          32,
          105,
          102,
          32,
          121,
          111,
          117,
          32,
          114,
          101,
          98,
          117,
          105,
          108,
          100,
          13,
          10,
          102,
          114,
          111,
          109,
          32,
          115,
          111,
          117,
          114,
          99,
          101,
          115,
          32,
          98,
          121,
          32,
          114,
          117,
          110,
          110,
          105,
          110,
          103,
          32,
          97,
          108,
          108,
          46,
          98,
          97,
          115,
          104,
          32,
          40,
          115,
          101,
          101,
          32,
          100,
          111,
          99,
          47,
          105,
          110,
          115,
          116,
          97,
          108,
          108,
          46,
          104,
          116,
          109,
          108,
          41,
          46,
          13,
          10,
          89,
          111,
          117,
          32,
          115,
          104,
          111,
          117,
          108,
          100,
          32,
          97,
          108,
          115,
          111,
          32,
          97,
          100,
          100,
          32,
          116,
          104,
          101,
          32,
          71,
          111,
          32,
          98,
          105,
          110,
          97,
          114,
          121,
          32,
          100,
          105,
          114,
          101,
          99,
          116,
          111,
          114,
          121,
          32,
          36,
          71,
          79,
          82,
          79,
          79,
          84,
          47,
          98,
          105,
          110,
          13,
          10,
          116,
          111,
          32,
          121,
          111,
          117,
          114,
          32,
          115,
          104,
          101,
          108,
          108,
          39,
          115,
          32,
          112,
          97,
          116,
          104,
          46,
          13,
          10,
          13,
          10,
          70,
          111,
          114,
          32,
          101,
          120,
          97,
          109,
          112,
          108,
          101,
          44,
          32,
          105,
          102,
          32,
          121,
          111,
          117,
          32,
          101,
          120,
          116,
          114,
          97,
          99,
          116,
          101,
          100,
          32,
          116,
          104,
          101,
          32,
          116,
          97,
          114,
          32,
          102,
          105,
          108,
          101,
          32,
          105,
          110,
          116,
          111,
          32,
          36,
          72,
          79,
          77,
          69,
          47,
          103,
          111,
          44,
          32,
          121,
          111,
          117,
          32,
          109,
          105,
          103,
          104,
          116,
          13,
          10,
          112,
          117,
          116,
          32,
          116,
          104,
          101,
          32,
          102,
          111,
          108,
          108,
          111,
          119,
          105,
          110,
          103,
          32,
          105,
          110,
          32,
          121,
          111,
          117,
          114,
          32,
          46,
          112,
          114,
          111,
          102,
          105,
          108,
          101,
          58,
          13,
          10,
          13,
          10,
          32,
          32,
          32,
          32,
          101,
          120,
          112,
          111,
          114,
          116,
          32,
          71,
          79,
          82,
          79,
          79,
          84,
          61,
          36,
          72,
          79,
          77,
          69,
          47,
          103,
          111,
          13,
          10,
          32,
          32,
          32,
          32,
          101,
          120,
          112,
          111,
          114,
          116,
          32,
          80,
          65,
          84,
          72,
          61,
          36,
          80,
          65,
          84,
          72,
          58,
          36,
          71,
          79,
          82,
          79,
          79,
          84,
          47,
          98,
          105,
          110,
          13,
          10,
          13,
          10,
          83,
          101,
          101,
          32,
          100,
          111,
          99,
          47,
          105,
          110,
          115,
          116,
          97,
          108,
          108,
          46,
          104,
          116,
          109,
          108,
          32,
          102,
          111,
          114,
          32,
          109,
          111,
          114,
          101,
          32,
          100,
          101,
          116,
          97,
          105,
          108,
          115
        ],
        "dir_disk_nbr": 0,
        "dir_records_this_disk": 1,
        "directory_offset": 662,
        "directory_records": 1,
        "directory_size": 76,
        "disk_nbr": 0
      },
      "offset": 738
    },
    "dir64": null,
    "global_offset": 0
  },
  "size": 1886
}
